### PR TITLE
Type.IsSZArray property.

### LINKED ISSUE #16566: Type: Needs api to distinguish SzArrays from Multidim arrays of rank 1.
The SZ array type ("T[]") and rank 1 multidim arrays ("T[*]") are distinct and non-interchangable types in .NET. However, the Reflection surface area contains no api to distinguish. (The most "obvious" approach ("t is Array && t.GetArrayRank() == 1") captures both.) 

Proposal:

   Add a new property to System.Type:
```c#
      public virtual bool IsSZArray { get { throw NotImplemented.ByDesign; }} // Subclass needs to implement
```
Alternate names that have been proposed:
       IsSZArray
       IsVector

The name "IsSzArray" has the small advantage that it already exists in both CoreClr and CoreRt - it's just not public.

### PR DESCRIPTION
**Pull request by JonHanna**
Closes #16566

Won't pass CI until dotnet/coreclr#10105 and corert implementation (pending) are in.

### CODE REVIEW

File: src/System.Runtime/tests/System/TypeTests.cs [hunk 1]
```diff
@@ -293,5 +293,29 @@ public static void ReflectionOnlyGetType()
...
+
+        private static IEnumerable<object[]> SZArrayOrNotTypes()
+        {
+            yield return new object[] {typeof(int[]), true};
+            yield return new object[] {typeof(string[]), true};
+            yield return new object[] {typeof(void), false};
+            yield return new object[] {typeof(int), false};
+            yield return new object[] {typeof(int[]).MakeByRefType(), false};
+            yield return new object[] {typeof(int[,]), false};
+            yield return new object[]{typeof(TypeTests), false };
```
Comment (AtsushiKan) [inline src/System.Runtime/tests/System/TypeTests.cs]: Nit: missing space between "]" and "{"

File: src/System.Runtime/tests/System/TypeTests.cs [hunk 2]
```diff
@@ -293,5 +293,29 @@ public static void ReflectionOnlyGetType()
...
+            yield return new object[] {Array.CreateInstance(typeof(int), new[] {2}, new[] {-1}).GetType(), false};
+            yield return new object[] {Array.CreateInstance(typeof(int), new[] {2}, new[] {1}).GetType(), false};
+            yield return new object[] {Array.CreateInstance(typeof(int), new[] {2}, new[] {0}).GetType(), true};
+            yield return new object[] {typeof(int[][]), true};
+            yield return new object[] { Type.GetType("System.Int32[]"), true };
+            yield return new object[] { Type.GetType("System.Int32[*]"), false };
```
Comment (AtsushiKan) [inline src/System.Runtime/tests/System/TypeTests.cs]: typeof(int).MakeArrayType() and typeof(int).MakeArrayType(1) are a less roundable way to get these types than calling Type.GetType().

Comment (AtsushiKan) [inline src/System.Runtime/tests/System/TypeTests.cs]: typeof(int).MakeArrayType() and typeof(int).MakeArrayType(1) are a less roundabout way to get these types than calling Type.GetType().

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Runtime/tests/System/TypeTests.cs (hunk 2 above)
Comment (JonHanna) [inline src/System.Runtime/tests/System/TypeTests.cs]: My intention had been to arive at types in a few different ways even if they're the same implementation underneath, so I've added the `MakeArrayType()` approach, but left these in.

**Comment by JonHanna** [comment]
That netfx failure is definitely related. Do I need to do something to make it work there, or do something to make the test not run there?

**Comment by stephentoub** [comment]
> Do I need to do something to make it work there, or do something to make the test not run there?

I expect you need to guard the tests to only be compiled when targeting a build with the new surface area:
https://github.com/JonHanna/corefx/blob/998ec2696819fae352e8a03d08922bda741a5ecf/src/System.Runtime/tests/System.Runtime.Tests.csproj#L125

File: src/System.Reflection.Emit/tests/EnumBuilder/EnumBuilder.Properties.Tests.cs [hunk 3]
```diff
@@ -30,5 +30,25 @@ public void Namespace()
...
+            Type arrType = enumBuilder.MakeArrayType();
+            Assert.True(arrType.IsArray);
+            Assert.True(arrType.IsSZArray);
+            arrType = enumBuilder.MakeArrayType(1);
+            Assert.True(arrType.IsArray);
+            Assert.False(arrType.IsSZArray);
+            arrType = enumBuilder.MakeArrayType(2);
+            Assert.True(arrType.IsArray);
+            Assert.False(arrType.IsSZArray);
+        }
```
Comment (stephentoub) [inline src/System.Reflection.Emit/tests/EnumBuilder/EnumBuilder.Properties.Tests.cs]: Nit: in all of these tests, can you add blank lines between the relevant sets of tests, e.g.
```C#
Type asType = enumBuilder.AsType();
Assert.False(asType.IsArray);
Assert.False(asType.IsSZArray);

Type arrType = enumBuilder.MakeArrayType();
Assert.True(arrType.IsArray);
Assert.True(arrType.IsSZArray);
```
I find it hard to read when it's all smushed together like this.

Comment (JonHanna) [inline src/System.Reflection.Emit/tests/EnumBuilder/EnumBuilder.Properties.Tests.cs]: It does suddenly look like a big block once you point that out! Changed.

Comment (stephentoub) [inline src/System.Reflection.Emit/tests/EnumBuilder/EnumBuilder.Properties.Tests.cs]: :smile:

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please

**Pull request merged**